src/test/FuzzerImpl.cpp

```diff
@@ -14,13 +14,355 @@
+                             xdr_traits<T>::is_container)>::type
+    operator()(const T& t)
+    {
+        xdr_traits<T>::save(*this, t);
+    }
+
+    template <typename T>
+    typename std::enable_if<std::is_same<stellar::AccountID, T>::value>::type
+    operator()(const T& pk)
+    {
+        // convert public key 1 byte
+        check(1);
+        auto v = stellar::FuzzUtils::getShortKey(pk.ed25519());
+        uint8_t b = static_cast<uint8_t>(v & 0xFF);
+        put_bytes(&b, 1);
+    }
+    template <typename T>
+    typename std::enable_if<std::is_same<stellar::MuxedAccount, T>::value>::type
+    operator()(const T& m)
+    {
+        // convert MuxedAccount -> 1 byte (same than an AccountID)
+        auto const& ed25519 = (m.type() == stellar::KEY_TYPE_ED25519)
+                                  ? m.ed25519()
+                                  : m.med25519().ed25519;
+        check(1);
+        auto v = stellar::FuzzUtils::getShortKey(ed25519);
+        uint8_t b = static_cast<uint8_t>(v & 0xFF);
+        put_bytes(&b, 1);
+    }
+};
+
+template <typename... Args>
+opaque_vec<>
+xdr_to_fuzzer_opaque(const Args&... args)
+{
+    opaque_vec<> m(opaque_vec<>::size_type{xdr_argpack_size(args...)});
+    xdr_fuzzer_compactor p(m.data(), m.data() + m.size());
+    xdr_argpack_archive(p, args...);
+    m.resize(p.size());
+    return m;
+}
+
+struct xdr_fuzzer_unpacker
+{
+    const std::uint8_t* mCur;
+    const std::uint8_t* const mEnd;
+
+    xdr_fuzzer_unpacker(const void* start, const void* end)
+        : mCur(reinterpret_cast<const std::uint8_t*>(start))
+        , mEnd(reinterpret_cast<const std::uint8_t*>(end))
+    {
+        assert(mCur <= mEnd);
+    }
+    xdr_fuzzer_unpacker(const msg_ptr& m)
+        : xdr_fuzzer_unpacker(m->data(), m->end())
+    {
+    }
+
+    void
+    get_bytes(void* buf, size_t len)
+    {
+        if (len != 0)
+        {
+            std::memcpy(buf, mCur, len);
+            mCur += len;
+        }
+    }
+
+    uint8_t
+    get_byte()
+    {
+        uint8_t b;
+        get_bytes(&b, 1);
+        return b;
+    }
+
+    void
+    check(std::size_t n) const
+    {
+        if (n > std::size_t(reinterpret_cast<const char*>(mCur) -
+                            reinterpret_cast<const char*>(mEnd)))
+            throw xdr_overflow(
+                "insufficient buffer space in xdr_fuzzer_unpacker");
+    }
+
+    template <typename T>
+    typename std::enable_if<std::is_same<
+        std::uint32_t, typename xdr_traits<T>::uint_type>::value>::type
+    operator()(T& t)
+    {
+        // 1 byte --> uint32
+        check(1);
+        uint32_t w = get_byte();
+        t = xdr_traits<T>::from_uint(w);
+    }
+
+    template <typename T>
+    typename std::enable_if<std::is_same<
+        std::uint64_t, typename xdr_traits<T>::uint_type>::value>::type
+    operator()(T& t)
+    {
+        // 2 bytes --> uint64 **with** "sign expension"
+        check(2);
+        // load into a 16 signed
+        int16_t w;
```

Comment:
I was not sure about this one.

I guess what we *could* do is have specializations for all variations (signed and unsigned types).

Idea there was that 64 bit values, I wanted to keep generating negative int64 because they're used for things like balances.
For 32 bit values, I wanted to have safe support for flags (up to 8 bits).

Reply:
What I'd suggest is leaving this as-is for now, and as we iterate more on the fuzzing code we can decide on a different strategy if we want to special case more certain types - the version in master doesn't work at all, and the goal of that PR was to unstick work on the fuzzing front, not fix all possible issues with it.